sliding window ops in c++

mywoodstock opened the pull request.

tt-nshanker committed:
#7083: New conv api refactoring (intermediate)
#7083: resnet50 WH batch size 20 works

mywoodstock committed:
#7083: halo config generation in cpp done, need testing
#7083: Halo op using new c++ config generation
#7083: new micro op function for halo to be used in place of utwhv2
#7083: New Halo struct which stores the sliding window config, and creates hash along with it
#7083: config tensors as part of program
#7083: composite conv updates + pybind

tt-nshanker committed:
#7083: wip composite conv cpp refactoring
#7083: halo config generation in cpp done, need testing
#7083: Halo op using new c++ config generation
#7083: wip micro conv op

tt-nshanker committed:
#0: adding composite conv op and new halo op to makefile wip

mywoodstock committed:
#7083: config tensor generation before create_program, move to device in create_program
#7083: clean compile

tt-nshanker committed:
#7083: swap with new c++ conv

mywoodstock committed:
#7083: maxpool composite in c++
#7083: maxpool2d c++ macro-op pybind
#7083: new maxpool unit tests

mywoodstock committed:
#0: resize instead of reserve!

mywoodstock committed:
#0: some fixes

arakhmati committed:
#0: updated to_weight_special_padding_tile_layout to use borrowed or owned storage

mywoodstock committed:
#7083: remote config fix
#7083: config identical to orig version
#7083: fix broken height sharded
#7083: height sharded tests pass on GS

sankarmanoj-tt committed:
#7083: Build with CMake works!

arakhmati committed:
#5389: added host buffer wrapper

sankarmanoj-tt committed:
#0: Split conv and matmul into cpp+hpp

sankarmanoj-tt committed:
#0: Fixed BFloat8_B for conv

sankarmanoj-tt committed:
#7083: Using ToDtype to convert BFloat8_B Bias
#7083: snap to tile fix

tt-nshanker committed:
#7083: turned on all height sharded and block sharded tests
#7083: fixes for bias conversion. Enabled all gs unit tests. One failing only
#7083: fixes for first conv and added explicit debug mode for conv
#7083: remove dtype, cleaning

mywoodstock committed:
#0: disable non-RN50 tests

tt-nshanker committed:
#7083: 1x1 convfixes

mywoodstock committed:
#0: rebase to main, fix compile issues

mywoodstock committed:
#0: disable small rn50 block test, needs to be tested with new api

mywoodstock committed:
#7083: RN50 debug ongoing
#7083: eltwise, handle bfp8 and others

mywoodstock committed:
#0: disable new version of maxpool unit tests, enable after debug

mywoodstock committed:
#0: b=1 not working for sd

sankarmanoj-tt committed:
#7083: Removed import models.utility_function in conv2d.py
#7083: Disabled failing tests on test_sd_conv

sankarmanoj-tt committed:
#0: Added ResnetBlock for SD Tests

mywoodstock committed:
#0: fixing some bugs. RN50 runs e2e

arakhmati commented on ttnn/cpp/ttnn/operations/maxpool2d.hpp:
```diff
@@ -0,0 +1,59 @@
+#include "tt_eager/tt_dnn/op_library/sliding_window_op_infra/halo_op.hpp"
+#include "tt_eager/tt_dnn/op_library/sliding_window_op_infra/sliding_window.hpp"
+
+using namespace tt;
```
Remove `using namespace tt;` from here

arakhmati commented on ttnn/cpp/ttnn/operations/conv2d.hpp:
```diff
@@ -0,0 +1,150 @@
+#include "tt_eager/tt_dnn/op_library/sliding_window_op_infra/sliding_window.hpp"
+#include "tt_eager/tt_dnn/op_library/sliding_window_op_infra/halo_op.hpp"
+
+using namespace tt;
```
Remove `using namespace tt;` from here

arakhmati commented on ttnn/cpp/ttnn/operations/maxpool2d.hpp:
```diff
@@ -0,0 +1,59 @@
+using namespace tt;
+
+namespace ttnn::operations {
+namespace maxpool {
```
keep the namespace same as the file name

arakhmati commented on ttnn/ttnn/device.py:
```diff
@@ -72,4 +73,14 @@ def dump_device_memory_state(device, prefix=""):
     ttl.device.DumpDeviceMemoryState(device, prefix)
 
 
+def is_wormhole_b0():
```
please change this function to take in `device` and figure out the arch from the device. Because what would happen if we have a heterogeneous system?

arakhmati commented on ttnn/ttnn/device.py:
```diff
@@ -72,4 +73,14 @@ def dump_device_memory_state(device, prefix=""):
+    return "wormhole_b0" in ARCH_NAME
+
+
+def is_grayskull():
```
Same here. Make this function to take in device and figure out the arch from the device.

arakhmati commented on ttnn/ttnn/operations/conv2d.py:
```diff
@@ -4,12 +4,26 @@
 
 from typing import Tuple, Union, Dict, Optional
 import torch
```
don't import torch on top-level. Import it wherever only wherever you use it

arakhmati commented on ttnn/ttnn/__init__.py:
```diff
@@ -464,7 +465,7 @@ def manage_config(name, value):
 from ttnn.operations import transformer
 from ttnn.operations import kv_cache
-from ttnn.operations.conv2d import Conv2d
+from ttnn.operations.conv2d import Conv2d, conv2d, ConvConfig
```
Please rename `ConvConfig` to `Conv2dConfig`

tt-aho commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -1242,6 +1242,24 @@ void HWCommandQueue::enqueue_command(T& command, bool blocking) {
     }
 }
 
+// TODO: Currently converting page ordering from interleaved to sharded and then doing contiguous read/write
+//  Look into modifying command to do read/write of a page at a time to avoid doing copy
```
Why is this added back?

tt-aho commented on tt_metal/impl/program/program.hpp:
```diff
@@ -212,6 +216,9 @@ class Program {
 
     void add_semaphore(const CoreRangeSet & crs, uint32_t address, uint32_t init_value, CoreType core_type=CoreType::WORKER);
 
+    friend void detail::AddConfigTensor(Program &program, const Tensor& config_tensor);
+    void add_config_tensor(const Tensor& config_tensor);
+
```
Isn't tensor an eager object and we want eager/metal to be separate?
Should we track buffers here instead?

mywoodstock replied:
Yeah needed access to `Tensor`. You mean instead add the buffer pointer?

mywoodstock replied:
Let me try with buffer ptr instead.

mywoodstock replied:
Previously we had a model level cache for configs, which we got rid of since we want it to be hidden from model user, so we had decided to add it to the program itself.

abhullar-tt replied:
is there some intermediate level where this info can be kept with the op?

tt-aho replied:
This seems like it should belong with / leverage the program cache infra? Since you are associating a program with specific tensor buffers

tt-aho commented on tt_eager/tt_dnn/op_library/sliding_window_op_infra/utils.hpp:
```diff
@@ -11,6 +11,10 @@ namespace tt::tt_metal {
 
 namespace utils {
 
+inline uint32_t nearest_y(uint32_t val, uint32_t divisor) {
+    return (val + divisor - 1) / divisor * divisor;
+}
+
```
This is a duplicate of `round_up`?

mywoodstock replied:
yes :) replaced

tt-aho commented on tt_eager/tt_dnn/op_library/sliding_window_op_infra/untilize_with_halo_config_generation_and_validation.py:
```diff
@@ -158,6 +158,8 @@ def decompose_conv_into_shards_and_generate_tensor_metadata(
         )
         conv_output_start_stick += conv_output_shard_height
 
+    # print("req_conv_input_shard_start_end=", req_conv_input_shard_start_end)
+
```
Remove dead code

tt-aho commented on tt_eager/tt_dnn/op_library/sliding_window_op_infra/tt_py_untilize_with_halo.py:
```diff
@@ -201,6 +201,10 @@ def core_id_to_physical_coord(core_id):
                 remote_read=remote_read,
             )
 
+            # print ("padding_config", padding_config)
+            # print ("local_config", local_config)
```
Remove dead code

mywoodstock replied:
removed

tt-aho reviewed (changes requested):
I think the big thing is that this pr is trying to add an eager dependency to metal with tensors. Don't really think that is something we want

pgkeller reviewed (approved):
don't know anything about most of this but no concerns about the small part that overlaps w/ me be codeowner...

mywoodstock committed:
#7083: Use device buffer instead of tensor to remove dependency from tt_metal, several other minor updates

mywoodstock commented:
> I think the big thing is that this pr is trying to add an eager dependency to metal with tensors. Don't really think that is something we want

@tt-aho I changed to use the device buffer instead of tensor and removed the tensor dependency from tt_metal. 
I think this should be good for now, if we decide to develop some infra for this on tt_eager side, we can move it then.

mywoodstock committed:
#7083: leftover dependency removal

tt-aho commented on tt_metal/impl/program/program.cpp:
```diff
@@ -516,6 +516,11 @@ void Program::add_semaphore(const CoreRangeSet & crs, uint32_t address, uint32_t
     semaphores_.emplace_back(Semaphore( crs, address, init_value, core_type));
 }
 
+void Program::add_config_buffer(std::shared_ptr<Buffer> config_buffer) {
+    this->invalidate_compile();
```
Why does adding the config buffer invalidate compile? Program is compiled with no knowledge of this.

mywoodstock replied:
Hmm, if the op config (buffer) is different for the same program config, we want to distinguish it. But thinking about it, currently we do not have different config for same program config. So I can remove this.

tt-aho reviewed (commented):
@abhullar-tt are you okay with this change for tracking these op config buffers in program object?

mywoodstock committed:
#7083: no need to invalidate compile

mywoodstock commented:
Nightly: https://github.com/tenstorrent/tt-metal/actions/runs/9182647129
Model perf: https://github.com/tenstorrent/tt-metal/actions/runs/9182651200

mywoodstock commented:
@tt-aho @abhullar-tt Any further objections on the program change?

mywoodstock committed:
#7083: resolved rn50 b=20 l1 issue. bad pcc with b=20, rest all good

mywoodstock commented:
model perf: https://github.com/tenstorrent/tt-metal/actions/runs/9199932628
